### PR TITLE
fix(codegen,runtime): Integer() raises ArgumentError on unparseable input

### PR DESCRIPTION
**Pull request by ryanseys**
## Summary

CRuby raises `ArgumentError` on `Integer("xyz")`, `Integer("")`, `Integer("12abc")`, etc. Spinel previously emitted a bare `(mrb_int)strtoll(s, NULL, 10)` which silently returned 0 on any unparseable input — meaning the natural CRuby idiom

```ruby
def parse_int(s) = Integer(s) rescue 0
```

never actually triggered the rescue path. `Integer("abc")` returned 0 directly without raising; the rescue was effectively dead code masking silent type-coercion bugs.

### Runtime (`lib/sp_runtime.h`)
- New `sp_str_to_i_strict` helper that wraps `strtoll` with proper endptr / whitespace handling. Raises `ArgumentError` with a CRuby-shaped message (`"invalid value for Integer(): \"...\""`) for nil, empty / whitespace-only, no-leading-digits, or trailing-junk inputs. Whitespace skipping uses `isspace((unsigned char)*p)` so it covers `\v` and `\f` in addition to space / tab / newline / CR.
- Forward declarations for `sp_raise_cls` and `sp_sprintf` so the new helper can call them.
- New `sp_mark_in_flight_exceptions` called by `sp_re_mark_globals`. `sp_str_to_i_strict` is the first runtime helper to hand a dynamically-allocated message (built via `sp_sprintf`, on the str-heap with mark byte 0xfe) to `sp_raise_cls`. That pointer lands in `sp_exc_msg[sp_exc_top-1]` and lives until the rescue handler reads it. Without marking, a GC cycle between the longjmp and the handler would sweep the message and leave the slot dangling.

### Codegen (`spinel_codegen.rb`)
- `compile_no_recv_call_expr` Integer() arm: routes the four string-typed dispatch sites (bare string arg, argv arg, and the two OrNode short-circuit branches) through `sp_str_to_i_strict`.
- `@needs_setjmp = 1` so any function using `Integer()` gets the setjmp infrastructure for the rescue catch path.

## Out of scope

- **Hex prefixes** (`Integer("0x10")` → 16 in CRuby) and **underscore separators** (`Integer("12_345")` → 12345 in CRuby). Spinel's base-10 `strtoll` path doesn't parse these; truly handling them needs either a custom parser or routing through `Integer(s, base)` explicit-base support.
- **Bigint auto-promotion** for over-`LLONG_MAX` inputs. CRuby silently promotes (`Integer("99999999999999999999999")` returns the actual value); Spinel currently saturates at `LLONG_MAX`. Strict `ERANGE → ArgumentError` would be over-strict vs CRuby; true parity needs Bignum integration in this path.
- **Mismatched-type rescue** (`Integer(s) rescue "default"` int+string) — needs poly-boxing fix.

## Test plan

- [x] `make` — builds clean
- [x] `make bootstrap` — gen2.c == gen3.c (bootstrap OK)
- [x] `make test` — full suite passes
- [x] `test/integer_argument_error.rb` covers:
  - Valid inputs (positive, zero, negative, +signed, all `isspace` whitespace classes including `\v` / `\f`)
  - Sign edge cases (`+0`, `-0`, sign with internal whitespace, double-sign, bare sign)
  - Edge magnitudes within `mrb_int` (LLONG_MAX, LLONG_MIN)
  - Invalid inputs (xyz, "", "   ", 12abc, 12.5)
  - Trailing-junk variants (42x, 42 x, 42\\n garbage, 42 trailing-ws)
  - Leading-junk variants (x42, " x 42")
  - Bare rescue with the formatted error message
  - **100-raise stress loop** exercising the GC-root fix — pre-fix this would non-deterministically read garbage from `sp_exc_msg`; post-fix all 100 raises survive with intact messages
- [x] `test/endless_method_rescue.rb` (pre-existing) continues to pass — `Integer("abc")` now raises but the rescue catches it; same observable output

### CODE REVIEW

File: lib/sp_runtime.h [hunk 1]
```diff
@@ -96,6 +96,30 @@ static mrb_int sp_int_clamp(mrb_int v,mrb_int lo,mrb_int hi){return v<lo?lo:v>hi
...
+   `(mrb_int)strtoll(s, NULL, 10)` spinel previously emitted silently
+   returned 0 instead, which made `Integer(s) rescue 0` always take
+   the main branch. This helper matches CRuby semantics: skips
+   leading/trailing whitespace, requires at least one valid digit,
+   rejects trailing junk. Accepts an optional leading `+` / `-`. */
+static mrb_int sp_str_to_i_strict(const char *s) {
+  if (!s) sp_raise_cls("ArgumentError", "invalid value for Integer(): nil");
+  const char *p = s;
+  while (*p == ' ' || *p == '\t' || *p == '\n' || *p == '\r') p++;
+  if (*p == '\0') sp_raise_cls("ArgumentError", sp_sprintf("invalid value for Integer(): \"%s\"", s));
```
[the remark this answers is not in the document]
Comment (ryanseys) [reply lib/sp_runtime.h:115]: Real bug, addressed in `a45da9a`.

You're right that this PR is the first to hand a dynamically-allocated message to `sp_raise_cls` — most existing raises pass string literals (rodata, marker byte ≠ 0xfe, so `sp_mark_string` is a safe no-op for them). `sp_str_to_i_strict`'s `sp_sprintf("invalid value for Integer(): \"%s\"", s)` produces a str-heap allocation with mark byte 0xfe; without GC-root marking, a sweep landing between the longjmp and the rescue handler reading the message would reap it.

Fix:

```c
static void sp_mark_in_flight_exceptions(void) {
  for (int i = 0; i < sp_exc_top; i++) sp_mark_string(sp_exc_msg[i]);
}
```

Called from `sp_re_mark_globals` (added a forward decl above its definition since the actual sp_exc_msg / sp_exc_top live further down the file with the rest of the exception infrastructure). `sp_mark_string` is null-safe and no-ops on non-heap pointers, so unconditional iteration over `[0, sp_exc_top)` is safe.

Test: added a 100-raise stress loop in `test/integer_argument_error.rb` that exercises the path. Pre-fix this would non-deterministically read garbage from `sp_exc_msg`; post-fix all 100 raises survive with intact messages.

File: lib/sp_runtime.h [hunk 2]
```diff
@@ -96,6 +96,30 @@ static mrb_int sp_int_clamp(mrb_int v,mrb_int lo,mrb_int hi){return v<lo?lo:v>hi
...
+   (empty string, leading/trailing junk, all-whitespace). The bare
+   `(mrb_int)strtoll(s, NULL, 10)` spinel previously emitted silently
+   returned 0 instead, which made `Integer(s) rescue 0` always take
+   the main branch. This helper matches CRuby semantics: skips
+   leading/trailing whitespace, requires at least one valid digit,
+   rejects trailing junk. Accepts an optional leading `+` / `-`. */
+static mrb_int sp_str_to_i_strict(const char *s) {
+  if (!s) sp_raise_cls("ArgumentError", "invalid value for Integer(): nil");
+  const char *p = s;
+  while (*p == ' ' || *p == '\t' || *p == '\n' || *p == '\r') p++;
```
[the remark this answers is not in the document]
Comment (ryanseys) [reply lib/sp_runtime.h]: Applied in `a45da9a`. Both leading and trailing whitespace skips now use `isspace((unsigned char)*p)`. `<ctype.h>` is already included in `lib/sp_runtime.h` (line 23). Test fixture extended to exercise `\v` and `\f` (which the previous manual check missed) — all CRuby-recognized whitespace now passes through correctly.

File: lib/sp_runtime.h [hunk 3]
```diff
@@ -96,6 +96,30 @@ static mrb_int sp_int_clamp(mrb_int v,mrb_int lo,mrb_int hi){return v<lo?lo:v>hi
...
+  if (*p == '\0') sp_raise_cls("ArgumentError", sp_sprintf("invalid value for Integer(): \"%s\"", s));
+  char *endptr;
+  long long v = strtoll(p, &endptr, 10);
```
[the remark this answers is not in the document]
Comment (ryanseys) [reply lib/sp_runtime.h:117]: Real concern but **out of scope** for this PR.

CRuby's actual behavior for `Integer("99999999999999999999999")` isn't `ArgumentError` — it silently promotes to `Bignum` and returns the value:

\`\`\`
\$ ruby -e 'puts Integer("99999999999999999999999")'
99999999999999999999999
\`\`\`

So strict `ERANGE → ArgumentError` would be over-strict vs CRuby. True parity requires Bignum auto-promotion in the `Integer(string)` path, which means routing through `sp_bigint`'s string parser — a broader runtime change than this PR.

The current behavior (silent saturation at `LLONG_MAX` / `LLONG_MIN`) is a pre-existing data-loss gap I've documented in the updated PR body's "Out of scope" section. Worth a follow-up that wires Bignum into the path; flagging here so we don't lose track.

File: lib/sp_runtime.h [hunk 4]
```diff
@@ -96,6 +96,30 @@ static mrb_int sp_int_clamp(mrb_int v,mrb_int lo,mrb_int hi){return v<lo?lo:v>hi
...
+  if (endptr == p) sp_raise_cls("ArgumentError", sp_sprintf("invalid value for Integer(): \"%s\"", s));
+  while (*endptr == ' ' || *endptr == '\t' || *endptr == '\n' || *endptr == '\r') endptr++;
```
[the remark this answers is not in the document]
Comment (ryanseys) [reply lib/sp_runtime.h]: Applied in `a45da9a` along with the leading-whitespace fix. Both sites now use `isspace((unsigned char)*p)`. The `(unsigned char)` cast is intentional — `isspace` has undefined behavior on `EOF`-equivalent negative `signed char` values, and forcing through `unsigned char` first avoids that.
